src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -0,0 +1,167 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <boost/intrusive/set.hpp>
+
+#include "common/ceph_time.h"
+
+#include "crimson/os/seastore/cached_extent.h"
+#include "crimson/os/seastore/journal.h"
+#include "crimson/os/seastore/seastore_types.h"
+#include "crimson/os/seastore/segment_manager.h"
+
+namespace crimson::os::seastore {
+class Transaction;
+
+class SegmentCleaner : public JournalSegmentProvider {
+public:
+  /// Config
+  struct config_t {
+    size_t num_segments = 0;
+    size_t segment_size = 0;
+    size_t target_journal_segments = 0;
+    size_t max_journal_segments = 0;
+
+    static config_t default_from_segment_manager(
+      SegmentManager &manager) {
+      return config_t{
+	manager.get_num_segments(),
+	static_cast<size_t>(manager.get_segment_size()),
+	2,
+	4};
+    }
+  };
+
+  /// Callback interface for querying and operating on segments
+  class ExtentCallbackInterface {
+  public:
+    /**
+     * get_next_dirty_extent
+     *
+     * returns all extents with dirty_from < bound
+     */
+    using get_next_dirty_extents_ertr = crimson::errorator<>;
+    using get_next_dirty_extents_ret = get_next_dirty_extents_ertr::future<
+      std::vector<CachedExtentRef>>;
+    virtual get_next_dirty_extents_ret get_next_dirty_extents(
+      journal_seq_t bound ///< [in] return extents with dirty_from < bound
+    ) = 0;
+
+    /**
+     * rewrite_extent
+     *
+     * Updates t with operations moving the passed extents to a new
+     * segment.  Segments in extents_to_relocate may be invalid,
```

Comment:
where is `extents_to_relocate` defined?

Reply:
Earlier iteration passed a list there, will fix.